Add guide: developing with Fusion across dbt Platform and local environments (#9047)
fishtown-analytics/docs.getdbt.com: website/docs/guides/fusion-platform-local-workflow.md

---
title: "Developing with Fusion: dbt Platform and local workflows"
id: "fusion-platform-local-workflow"
level: 'Intermediate'
icon: 'zap'

> hide_table_of_contents: false

Reviewer: ```suggestion
hide_table_of_contents: true
```

## Prerequisites

- You have a <Constant name="dbt_platform" /> account with at least one project using the <Constant name="fusion_engine" />.
- You have [installed Fusion locally](/docs/local/install-dbt?version=2#installation) — either via the CLI or the VS Code extension.

---

## 1. Managing credentials

How you authenticate to your data warehouse locally depends on which local tool you use.

### Fusion CLI (Cloud CLI)

If you're developing locally with only the CLI, Fusion works natively with the [dbt Cloud CLI](/docs/cloud/cloud-cli-installation). The Cloud CLI authenticates using your dbt Platform session. Your warehouse credentials are managed centrally in dbt Platform and passed through automatically.

```shell
# Install the Cloud CLI (if not already installed)
pip install dbt

# Verify your connection
dbt debug
```

This is the lowest-friction path for teams that don't need full IDE integration locally.

### VS Code extension (profiles.yml required)

The dbt VS Code extension runs Fusion, and its LSP, in a local process and connects directly to your warehouse. For this reason, a `profiles.yml` is required for local extension development sessions.

> To avoid manually recreating your warehouse configuration, use `dbt init`. If you have a `dbt_cloud.yml` in your project (created when you clone a dbt Platform project), Fusion will attempt to hydrate non-sensitive credential metadata from dbt Platform automatically:

Reviewer 2: Worth a call out here of what the flow looks like if they don't already have a 'dbt_cloud.yml' in their project? I think the 'dbt init' would walk them through question/input in the terminal that would then create the 'dbt_could.yml' for them right?

Reviewer 3: Added link to the instructions

### Fusion CLI (Cloud CLI)

When using the Cloud CLI, environment variables configured in dbt Platform are automatically available during your local development session — the same variables used in production are injected into your CLI session. No extra setup is needed.

### VS Code extension (.env file)

The VS Code extension runs Fusion as a local process, so platform-managed environment variables are not automatically available. Instead, use a [`.env` file](https://dotenvx.com/docs) at the root of your dbt project:

```shell
# .env
DBT_MY_DATABASE=my_database
DBT_MY_SCHEMA=my_dev_schema
DBT_TARGET_SCHEMA=analytics_dev
```

Fusion and the VS Code extension automatically load values from this file. You can also view and override individual environment variables from the VS Code extension's settings UI.

Reference these variables in your `profiles.yml` or elsewhere in your dbt project using the [`env_var` Jinja function](/reference/dbt-jinja-functions/env_var):

```yaml
# profiles.yml
my_profile:
  target: dev
  outputs:
    dev:
      type: snowflake
      account: my_account
      database: "{{ env_var('DBT_MY_DATABASE') }}"
      schema: "{{ env_var('DBT_MY_SCHEMA') }}"
```

> For a full walkthrough of `.env` file usage and variable precedence, see [Environment variables](/docs/local/install-dbt?version=2#environment-variables).

Reviewer 2: This might just be because of the "vercel.app" preview but this Environment variables link didn't seem to link to the correct docs page

Reviewer 4: thinking we can also link to https://docs.getdbt.com/docs/build/environment-variables to show people where in dbt platform they can see the env vars that are set for their project?

```shell
echo ".env" >> .gitignore
```
:::

When environment variables change in <Constant name="dbt_platform" /> (you add variables or rename values), update `.env.example` in the same pull request so local developers know to update their own `.env`.

:::tip For teams with strict security requirements 

Consider a script that fetches variables from your secrets manager (for example, AWS Secrets Manager or 1Password) and writes them to `.env` at the start of a session, instead of storing values in a file long term.

:::

## 3. Managing Fusion versions

The **Latest** release track on <Constant name="dbt_platform" /> updates continuously as <Constant name="fusion" /> ships new releases. If your local version falls behind, you might see inconsistent behavior. The same query could compile differently locally than in production, or a feature might exist in <Constant name="dbt_platform" /> but not in your local binary. Stay current to avoid these mismatches.

### Versions on the dbt platform

> On <Constant name="dbt_platform" />, <Constant name="<Constant name="fusion" /> follows a versionless release track model. The default release track is **Latest**, which always runs the most recent stable release. For details on release tracks and their stability levels, see [<Constant name="fusion" /> releases](/docs/fusion/fusion-releases).

Reviewer: ```suggestion
On <Constant name="dbt_platform" />, <Constant name="fusion" /> follows a versionless release track model. The default release track is **Latest**, which always runs the most recent stable release. For details on release tracks and their stability levels, see [<Constant name="fusion" /> releases](/docs/fusion/fusion-releases).
```

> On <Constant name="dbt_platform" />, <Constant name="<Constant name="fusion" />" /> follows a versionless release track model. The default release track is **Latest**, which always runs the most recent stable release. For details on release tracks and their stability levels, see [<Constant name="fusion" /> releases](/docs/fusion/fusion-releases).

Reviewer: ```suggestion
On <Constant name="dbt_platform" />, <Constant name="<Constant name="fusion" /> follows a versionless release track model. The default release track is **Latest**, which always runs the most recent stable release. For details on release tracks and their stability levels, see [<Constant name="fusion" /> releases](/docs/fusion/fusion-releases).
```

---
title: "Hybrid development with dbt platform and Fusion"
id: "fusion-platform-local-workflow"
level: 'Beginner'
icon: 'zap'
hide_table_of_contents: true
tags: ['dbt Fusion engine', 'dbt platform', 'Local development']
recently_updated: true
---

## Introduction

> Hybrid dbt deployments are becoming increasingly common. <Constant name="fusion" /> adopters are no different, frequently working in several places at once: in the <Constant name="dbt_platform" /> for production runs and IDE-based development, and locally using the dbt platform CLI or the dbt VS Code extension. 

Reviewer 3: ```suggestion
Hybrid dbt deployments are becoming increasingly common. <Constant name="fusion" /> adopters are frequently working in several places at once: in the <Constant name="dbt_platform" /> for production runs and IDE-based development, and locally using the dbt platform CLI or the dbt VS Code extension. 
```

### dbt platform CLI

When you use the dbt platform CLI, <Constant name="dbt_platform" /> injects the same environment variables you use in production into your dbt CLI session. You don't need extra setup.

### VS Code extension (.env file)

The dbt VS Code extension runs <Constant name="fusion" /> as a local process, so environment variables from <Constant name="dbt_platform" /> are not automatically available. Instead, use a [`.env` file](https://dotenvx.com/docs) at the root of your dbt project:

```shell
# .env
DBT_MY_DATABASE=my_database
DBT_MY_SCHEMA=my_dev_schema
DBT_TARGET_SCHEMA=analytics_dev
```

<Constant name="fusion" /> and the dbt VS Code extension automatically load values from this file. You can also view and override individual environment variables from the extension's settings UI.

Reference these variables in your `profiles.yml` or elsewhere in your dbt project using the [`env_var` Jinja function](/reference/dbt-jinja-functions/env_var):

```yaml
# profiles.yml
my_profile:
  target: dev
  outputs:
    dev:
      type: snowflake
      account: my_account
      database: "{{ env_var('DBT_MY_DATABASE') }}"
      schema: "{{ env_var('DBT_MY_SCHEMA') }}"
```

> For a full walkthrough of `.env` file usage and variable precedence, see [Environment variables](/docs/local/install-dbt?version=2#environment-variables).

Reviewer 3: ```suggestion
For a full walkthrough of `.env` file usage and variable precedence, see [Environment variables](/docs/build/environment-variables) for more information on where to find your configured variables and [Set environment variables locally](/docs/configure-dbt-extension?version=2.0#set-environment-variables-locally) for local configuration instructions.
```
